File: packages/core-node/src/providers/mtn-momo.ts
import { randomUUID, timingSafeEqual } from "node:crypto";

import {
  PaymentError,
  PaymentStatus,
  type PaymentEvent,
  type PaymentProvider,
  type PaymentRequest,
  type PaymentSession,
  type PaymentStatusResult,
  type RefundResult,
} from "../types.js";
import { validateRefundAmount } from "../refund-validation.js";
import { InMemoryWebhookEventStore, type WebhookEventStore } from "../webhook-event-store.js";

const BASE_URLS = {
  sandbox: "https://sandbox.momodeveloper.mtn.com",
  production: "https://proxy.momoapi.mtn.com",
} as const;

type MtnMomoEnvironment = keyof typeof BASE_URLS;

export interface MtnMomoProviderConfig {
  subscriptionKey: string;
  apiUser: string;
  apiKey: string;
  targetEnvironment: MtnMomoEnvironment;
  defaultCurrency: string;
  /** Override for local API-compatible mocks. */
  baseUrl?: string;
  webhookEventStore?: WebhookEventStore;
}

interface MtnTokenResponse {
  access_token?: string;
  expires_in?: number;
}

interface MtnRequestToPay {
  amount?: string;
  currency?: string;
  externalId?: string;
  referenceId?: string;
  status?: string;
  code?: string;
}

interface MtnErrorResponse {
  code?: string;
  message?: string;
}

interface MtnWebhookPayload extends MtnRequestToPay {
  id?: string;
  timestamp?: string;
}

export class MtnMomoProviderError extends Error {
  public constructor(public readonly code: PaymentError, message: string) {
    super(message);
    this.name = "MtnMomoProviderError";
  }
}

export class MtnMomoProvider implements PaymentProvider {
  private accessToken?: string;
  private accessTokenExpiresAt = 0;
  private readonly webhookEventStore: WebhookEventStore;

  public constructor(private readonly config: MtnMomoProviderConfig) {
    this.webhookEventStore = config.webhookEventStore ?? new InMemoryWebhookEventStore();
  }

  public async initiatePayment(params: PaymentRequest): Promise<PaymentSession> {
    if (params.customerPhone === undefined || params.customerPhone.length === 0) {
      throw new MtnMomoProviderError(PaymentError.InvalidPhone, "MTN MoMo requires customerPhone");
    }

    const sessionId = randomUUID();
    await this.request("/collection/v1_0/requesttopay", {
      method: "POST",
      headers: { "X-Reference-Id": sessionId },
      body: JSON.stringify({
        amount: params.amount.toString(),
        currency: params.currency || this.config.defaultCurrency,
        externalId: params.reference,
        payer: { partyIdType: "MSISDN", partyId: params.customerPhone },
        payerMessage: "Payment request",
        payeeNote: params.reference,
      }),
    });

    return {
      id: sessionId,
      reference: params.reference,
      amount: params.amount,
      currency: params.currency || this.config.defaultCurrency,
      status: PaymentStatus.Pending,
    };
  }

  public async checkStatus(sessionId: string): Promise<PaymentStatusResult> {
    const transaction = await this.getRequestToPay(sessionId);
    if (typeof transaction.status !== "string") {
      throw new MtnMomoProviderError(PaymentError.Unknown, "MTN MoMo response is missing status");
    }
    return this.toPaymentStatusResult(transaction.status, transaction.code);
  }

  public async handleWebhook(
    rawBody: string | Buffer,
    headers: Record<string, string | string[] | undefined>
  ): Promise<PaymentEvent> {
    const receivedKey = this.findHeader(headers, "ocp-apim-subscription-key");
    if (receivedKey === undefined || !this.isValidSubscriptionKey(receivedKey)) {
      throw new MtnMomoProviderError(PaymentError.Unknown, "Invalid MTN MoMo webhook security key");
    }

    const serializedBody = typeof rawBody === "string" ? rawBody : rawBody.toString("utf8");
    const payload = this.parseWebhookPayload(serializedBody);
    const sessionId = payload.referenceId;
    if (sessionId === undefined || sessionId.length === 0 || typeof payload.status !== "string") {
      throw new MtnMomoProviderError(PaymentError.Unknown, "Incomplete MTN MoMo webhook payload");
    }

    const status = this.toPaymentStatus(payload.status);
    const ref = payload.externalId === undefined ? {} : { reference: payload.externalId };
    const occurredAt = payload.timestamp ?? new Date().toISOString();
    const id = payload.id ?? sessionId;

    const event: PaymentEvent = status === PaymentStatus.Failed
      ? { id, sessionId, status: PaymentStatus.Failed, ...ref, occurredAt, error: this.mapError(200, payload.code) }
      : { id, sessionId, status, ...ref, occurredAt };
    return this.webhookEventStore.record(event);
  }

  public async refund(sessionId: string, amount?: number): Promise<RefundResult> {
    if (amount !== undefined) {
      validateRefundAmount(
        amount,
        (code, message) => new MtnMomoProviderError(code, message)
      );
    }

    const originalTransaction = await this.getRequestToPay(sessionId);
    const originalAmount = this.parseAmount(originalTransaction.amount);

    if (amount !== undefined && amount > originalAmount) {
      throw new MtnMomoProviderError(
        PaymentError.RefundAmountExceedsBalance,
        "Refund amount exceeds the original payment amount"
      );
    }

    const refundAmount = amount ?? originalAmount;
    const refundId = randomUUID();

    await this.request("/collection/v1_0/refund", {
      method: "POST",
      headers: { "X-Reference-Id": refundId },
      body: JSON.stringify({
        amount: refundAmount.toString(),
        currency: this.config.defaultCurrency,
        externalId: sessionId,
        payerMessage: "Refund",
        payeeNote: "Refund",
      }),
    });

    return {
      sessionId,
      refundId,
      amount: refundAmount,
      status: PaymentStatus.Pending,
    };
  }

  private async getRequestToPay(sessionId: string): Promise<MtnRequestToPay> {
    const response = await this.request(`/collection/v1_0/requesttopay/${encodeURIComponent(sessionId)}`);
    return this.readJson<MtnRequestToPay>(response);
  }

  private async request(path: string, init: RequestInit = {}): Promise<Response> {
    const token = await this.getAccessToken();
    let response: Response;
    try {
      response = await fetch(`${this.baseUrl()}${path}`, {
        ...init,
        headers: {
          Authorization: `Bearer ${token}`,
          "X-Target-Environment": this.config.targetEnvironment,
          "Ocp-Apim-Subscription-Key": this.config.subscriptionKey,
          "Content-Type": "application/json",
          ...init.headers,
        },
      });
    } catch (error: unknown) {
      throw this.toProviderError(error);
    }

    if (!response.ok) {
      throw await this.toResponseError(response);
    }
    return response;
  }

  private async getAccessToken(): Promise<string> {
    if (this.accessToken !== undefined && Date.now() < this.accessTokenExpiresAt) {
      return this.accessToken;
    }

    const credentials = Buffer.from(`${this.config.apiUser}:${this.config.apiKey}`).toString("base64");
    let response: Response;
    try {
      response = await fetch(`${this.baseUrl()}/collection/token/`, {
        method: "POST",
        headers: {
          Authorization: `Basic ${credentials}`,
          "Ocp-Apim-Subscription-Key": this.config.subscriptionKey,
        },
      });
    } catch (error: unknown) {
      throw this.toProviderError(error);
    }

    if (!response.ok) {
      throw await this.toResponseError(response);
    }
    const body = await this.readJson<MtnTokenResponse>(response);
    if (typeof body.access_token !== "string" || body.access_token.length === 0) {
      throw new MtnMomoProviderError(PaymentError.Unknown, "MTN MoMo token response is invalid");
    }

    this.accessToken = body.access_token;
    const expiresInSeconds = typeof body.expires_in === "number" ? body.expires_in : 300;
    this.accessTokenExpiresAt = Date.now() + Math.max(0, expiresInSeconds - 30) * 1_000;
    return this.accessToken;
  }

  private baseUrl(): string {
    return (this.config.baseUrl ?? BASE_URLS[this.config.targetEnvironment]).replace(/\/$/, "");
  }

  private async toResponseError(response: Response): Promise<MtnMomoProviderError> {
    const body = await this.readJsonOrEmpty<MtnErrorResponse>(response);
    const message = body.message ?? `MTN MoMo request failed with HTTP ${response.status}`;
    return new MtnMomoProviderError(this.mapError(response.status, body.code), message);
  }

  private toProviderError(error: unknown): MtnMomoProviderError {
    return new MtnMomoProviderError(
      PaymentError.ProviderTimeout,
      error instanceof Error ? error.message : "MTN MoMo request failed"
    );
  }

  private mapError(httpStatus: number, mtnCode: string | undefined): PaymentError {
    if (["RESOURCE_NOT_FOUND", "PAYER_NOT_FOUND", "NOT_ENOUGH_FUNDS"].includes(mtnCode ?? "")) {
      return PaymentError.InsufficientFunds;
    }
    if (["APPROVAL_REJECTED", "EXPIRED"].includes(mtnCode ?? "")) {
      return PaymentError.UserCancelled;
    }
    if (httpStatus === 408 || httpStatus === 429 || httpStatus >= 500) {
      return PaymentError.ProviderTimeout;
    }
    return PaymentError.Unknown;
  }

  private toPaymentStatus(status: string): PaymentStatus {
    switch (status.toUpperCase()) {
      case "SUCCESSFUL":
        return PaymentStatus.Success;
      case "PENDING":
        return PaymentStatus.Pending;
      case "FAILED":
        return PaymentStatus.Failed;
      default:
        throw new MtnMomoProviderError(PaymentError.Unknown, "Unknown MTN MoMo payment status");
    }
  }

  private toPaymentStatusResult(status: string, errorCode?: string): PaymentStatusResult {
    const paymentStatus = this.toPaymentStatus(status);
    if (paymentStatus !== PaymentStatus.Failed) {
      return { status: paymentStatus };
    }

    return { status: paymentStatus, error: this.mapError(200, errorCode) };
  }

  private parseAmount(value: string | undefined): number {
    const amount = value === undefined ? Number.NaN : Number(value);
    if (!Number.isSafeInteger(amount) || amount <= 0) {
      throw new MtnMomoProviderError(PaymentError.Unknown, "MTN MoMo response is missing original amount");
    }
    return amount;
  }

  private findHeader(
    headers: Record<string, string | string[] | undefined>,
    expectedName: string
  ): string | undefined {
    for (const [name, value] of Object.entries(headers)) {
      if (name.toLowerCase() !== expectedName) {
        continue;
      }
      if (typeof value === "string") {
        return value;
      }
      if (Array.isArray(value) && value.length === 1) {
        return value[0];
      }
    }
    return undefined;
  }

  private isValidSubscriptionKey(receivedKey: string): boolean {
    const expected = Buffer.from(this.config.subscriptionKey);
    const received = Buffer.from(receivedKey);
    return expected.length === received.length && timingSafeEqual(expected, received);
  }

  private parseWebhookPayload(serializedBody: string): MtnWebhookPayload {
    try {
      const payload: unknown = JSON.parse(serializedBody);
      if (typeof payload !== "object" || payload === null || Array.isArray(payload)) {
        throw new Error("Webhook payload must be an object");
      }
      return payload as MtnWebhookPayload;
    } catch {
      throw new MtnMomoProviderError(PaymentError.Unknown, "Invalid MTN MoMo webhook payload");
    }
  }

  private async readJson<T>(response: Response): Promise<T> {
    const body: unknown = await response.json();
    return body as T;
  }

  private async readJsonOrEmpty<T>(response: Response): Promise<T> {
    try {
      return await this.readJson<T>(response);
    } catch {
      return {} as T;
    }
  }
}

File: packages/core-python/waslpay/providers/mtn_momo.py
from __future__ import annotations

import base64
import json
import sys
import time
import uuid
from typing import Any

import httpx

from ..contracts import PaymentProvider, WebhookHeaders
from ..enums import PaymentError, PaymentStatus
from ..errors import ProviderError
from ..models import PaymentEvent, PaymentRequest, PaymentSession, PaymentStatusResult, RefundResult
from ..refund_validation import validate_refund_amount
from ..webhook_event_store import InMemoryWebhookEventStore, WebhookEventStore


class MtnMomoProvider(PaymentProvider):
    _BASE_URLS = {"sandbox": "https://sandbox.momodeveloper.mtn.com", "production": "https://proxy.momoapi.mtn.com"}

    def __init__(self, client: httpx.AsyncClient, subscription_key: str, api_user: str, api_key: str, target_environment: str = "sandbox", default_currency: str = "XOF", webhook_event_store: WebhookEventStore | None = None, base_url: str | None = None) -> None:
        self._client, self._subscription_key, self._api_user, self._api_key = client, subscription_key, api_user, api_key
        self._target_environment, self._default_currency, self._base_url = target_environment, default_currency, (base_url if base_url is not None else self._BASE_URLS[target_environment]).rstrip("/")
        self._token: str | None = None; self._token_expires_at = 0.0
        # A process-wide singleton would couple unrelated app instances; production must
        # inject durable storage shared by its webhook handling workers.
        self._webhook_event_store = (
            webhook_event_store
            if webhook_event_store is not None
            else InMemoryWebhookEventStore()
        )

    async def initiate_payment(self, params: PaymentRequest) -> PaymentSession:
        if not params.customer_phone: raise ProviderError(PaymentError.INVALID_PHONE, "MTN MoMo requires customer_phone")
        session_id = str(uuid.uuid4())
        await self._request("POST", "/collection/v1_0/requesttopay", session_id, json={"amount": str(params.amount), "currency": params.currency or self._default_currency, "externalId": params.reference, "payer": {"partyIdType": "MSISDN", "partyId": params.customer_phone}, "payerMessage": "Payment request", "payeeNote": params.reference})
        return PaymentSession(id=session_id, reference=params.reference, amount=params.amount, currency=params.currency or self._default_currency, status=PaymentStatus.PENDING)

    async def check_status(self, session_id: str) -> PaymentStatusResult:
        payload = await self._transaction(session_id)
        return self._status_result(payload.get("status"), payload.get("code"))

    async def handle_webhook(self, raw_body: str | bytes, headers: WebhookHeaders) -> PaymentEvent:
        if self._header(headers, "ocp-apim-subscription-key") != self._subscription_key: raise ProviderError(PaymentError.UNKNOWN, "Invalid MTN MoMo webhook security key")
        payload = self._decode(raw_body); session_id, status = payload.get("referenceId"), payload.get("status")
        if not isinstance(session_id, str): raise ProviderError(PaymentError.UNKNOWN, "Incomplete MTN MoMo webhook payload")
        payment_status = self._status(status)
        error = self._map_error(200, payload.get("code")) if payment_status is PaymentStatus.FAILED else None
        event = PaymentEvent(id=str(payload.get("id", session_id)), session_id=session_id, status=payment_status, occurred_at=str(payload.get("timestamp", "1970-01-01T00:00:00Z")), reference=payload.get("externalId") if isinstance(payload.get("externalId"), str) else None, error=error)
        return self._webhook_event_store.process(event, self._process_webhook_event)

    @staticmethod
    def _process_webhook_event(event: PaymentEvent) -> PaymentEvent:
        return event

    async def refund(self, session_id: str, amount: int | float | None = None) -> RefundResult:
        if amount is not None:
            amount = validate_refund_amount(
                amount,
                lambda code, message: ProviderError(code, message),
            )

        transaction = await self._transaction(session_id)
        original_amount = self._original_amount(transaction.get("amount"))

        if amount is not None and amount > original_amount:
            raise ProviderError(
                PaymentError.REFUND_AMOUNT_EXCEEDS_BALANCE,
                "Refund amount exceeds the original payment amount",
            )

        refund_amount = amount if amount is not None else original_amount
        refund_id = str(uuid.uuid4())
        await self._request("POST", "/collection/v1_0/refund", refund_id, json={"amount": str(refund_amount), "currency": self._default_currency, "externalId": session_id, "payerMessage": "Refund", "payeeNote": "Refund"})
        return RefundResult(session_id=session_id, refund_id=refund_id, amount=refund_amount, status=PaymentStatus.PENDING)

    async def _transaction(self, session_id: str) -> dict[str, Any]: return await self._request("GET", f"/collection/v1_0/requesttopay/{session_id}")
    async def _request(self, method: str, path: str, reference_id: str | None = None, **kwargs: Any) -> dict[str, Any]:
        token = await self._access_token(); headers = {"Authorization": f"Bearer {token}", "X-Target-Environment": self._target_environment, "Ocp-Apim-Subscription-Key": self._subscription_key}
        if reference_id: headers["X-Reference-Id"] = reference_id
        try: response = await self._client.request(method, self._base_url + path, headers=headers, **kwargs)
        except httpx.HTTPError as exc: raise ProviderError(PaymentError.PROVIDER_TIMEOUT, str(exc)) from exc
        return self._response(response)
    async def _access_token(self) -> str:
        if self._token and time.time() < self._token_expires_at: return self._token
        basic = base64.b64encode(f"{self._api_user}:{self._api_key}".encode()).decode(); response = await self._client.post(self._base_url + "/collection/token/", headers={"Authorization": f"Basic {basic}", "Ocp-Apim-Subscription-Key": self._subscription_key}); payload = self._response(response); token = payload.get("access_token")
        if not isinstance(token, str): raise ProviderError(PaymentError.UNKNOWN, "Invalid MTN MoMo token response")
        self._token, self._token_expires_at = token, time.time() + int(payload.get("expires_in", 300)) - 30; return token
    def _response(self, response: httpx.Response) -> dict[str, Any]:
        payload = response.json() if response.content else {}; payload = payload if isinstance(payload, dict) else {}
        if response.is_error:
            error = self._map_error(response.status_code, payload.get("code"))
            raise ProviderError(error, str(payload.get("message", "MTN MoMo request failed")))
        return payload
    @staticmethod
    def _status(status: object) -> PaymentStatus:
        payment_status = {
            "SUCCESSFUL": PaymentStatus.SUCCESS,
            "PENDING": PaymentStatus.PENDING,
            "FAILED": PaymentStatus.FAILED,
        }.get(str(status).upper())
        if payment_status is None:
            raise ProviderError(PaymentError.UNKNOWN, "Unknown MTN MoMo payment status")
        return payment_status
    @staticmethod
    def _original_amount(value: object) -> int:
        if isinstance(value, bool): raise ProviderError(PaymentError.UNKNOWN, "MTN MoMo response is missing original amount")
        if isinstance(value, int) and 0 < value <= sys.maxsize: return value
        if isinstance(value, str) and value.isdigit():
            parsed = int(value)
            if 0 < parsed <= sys.maxsize: return parsed
        raise ProviderError(PaymentError.UNKNOWN, "MTN MoMo response is missing original amount")
    def _status_result(self, status: object, error_code: object) -> PaymentStatusResult:
        payment_status = self._status(status)
        return PaymentStatusResult(status=payment_status, error=self._map_error(200, error_code) if payment_status is PaymentStatus.FAILED else None)
    @staticmethod
    def _map_error(http_status: int, code: object) -> PaymentError:
        normalized_code = str(code) if code is not None else ""
        if normalized_code in {"RESOURCE_NOT_FOUND", "PAYER_NOT_FOUND", "NOT_ENOUGH_FUNDS"}: return PaymentError.INSUFFICIENT_FUNDS
        if normalized_code in {"APPROVAL_REJECTED", "EXPIRED"}: return PaymentError.USER_CANCELLED
        if http_status >= 500 or http_status in {408, 429}: return PaymentError.PROVIDER_TIMEOUT
        return PaymentError.UNKNOWN
    @staticmethod
    def _decode(raw: str | bytes) -> dict[str, Any]:
        payload = json.loads(raw)
        if not isinstance(payload, dict): raise ProviderError(PaymentError.UNKNOWN, "Invalid MTN MoMo webhook payload")
        return payload
    @staticmethod
    def _header(headers: WebhookHeaders, name: str) -> str | None: return next((value for key, value in headers.items() if key.lower() == name and isinstance(value, str)), None)
